### PR TITLE
Feature/resolve images in helm chart values

### PR DESCRIPTION
**Pull request by nashton**

### CODE REVIEW

File: syncctl.py [hunk 1]
```diff
@@ -172,16 +172,7 @@ def template_flux(root_dir: str, git_repo: str, dir: str, git_repos: dict = coll
...
-    if not Path("work/flux").is_dir():
-        raise Exception('Please run mirror-flux before mirror-charts')
-    for dir in ["work/helm-chart-repo.tmp", "work/flux/flux/charts"]:
-        if Path(dir).is_dir():
-            shutil.rmtree(dir)
-        os.makedirs(dir)
-
-    manifests = template_flux("work/flux", config["flux_repository"].get("flux_repository", config["flux_repository"]["repository"]), "work/flux/" + config["flux_repository"]["entrypoint"])
-
+def get_helm_releases(config: dict, manifests: str) -> tuple[dict, list[dict]]:
```
Comment (jacobcrawford) [inline syncctl.py:175]: We should add a docsstring to the method that describes what is returned, and the input parameters.

File: syncctl.py [hunk 2]
```diff
@@ -216,6 +207,20 @@ def mirror_charts(config: dict, manifest: dict, manifest_file: str) -> None:
                 "values": values
             })
 
+    return helm_repos, helm_charts
+
+def mirror_charts(config: dict, manifest: dict, manifest_file: str) -> None:
```
Comment (jacobcrawford) [inline syncctl.py:212]: can we add docstrings here as well? Seems like there is a lot of implicit knowledge lost here.
Why does the function take both a manifest and a manifest file?
 Where are the charts mirrored and in what format?

File: syncctl.py [hunk 3]
```diff
@@ -290,19 +295,40 @@ def mirror_images(manifest: dict, manifest_file: str, incremental: bool) -> None
     os.rename("work/images.tmp", "work/images")
     save_manifest(manifest, manifest_file)
 
-def template_charts(api_versions: list[str], values: dict[str, str]) -> Generator[int, None, None]:
+def template_charts(api_versions: list[str], values: dict[str, str], helm_releases: list[dict], helm_charts: list[Path]) -> Generator[int, None, None]:
```
Comment (jacobcrawford) [inline syncctl.py:298]: This could also benefit from docstrings.

File: syncctl.py [hunk 4]
```diff
@@ -384,7 +410,10 @@ def resolve_images(config: dict, manifest: dict, manifest_file: str) -> None:
                 images[image] = process_image(image)
 
     manifests = template_flux("work/flux", config["flux_repository"].get("flux_repository", config["flux_repository"]["repository"]), "work/flux/" + config["flux_repository"]["entrypoint"])
-    for m in itertools.chain(iter([manifests]), template_charts(helm_config.get("api_versions", []), helm_config.get("values", {}))):
+
+    null, helm_releases = get_helm_releases(config, manifests)
```
Comment (jacobcrawford) [inline syncctl.py:414]: Is 'null' unused?
I think we should stick to the "_" for unused return variables. It is the most general syntax across different languages.

File: .github/workflows/pytest.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      - uses: actions/checkout@v3
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install pytest
+          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
```
Comment (jacobcrawford) [inline .github/workflows/pytest.yml:27]: Don't we always have a requirements file?
Why is this check necessary?

**Review by jacobcrawford — CHANGES_REQUESTED** [review]
Great with the tests! 
I think while the code is fresh in memory, we should add some docstrings to the functions.
It would help the readers, maintainers, and reviewers a lot :)

File: .github/workflows/pytest.yml (hunk 5 above)
Comment (nashton) [reply .github/workflows/pytest.yml:27]: No not in this repository, but if it is added later, the pipeline supports it.

File: syncctl.py (hunk 3 above)
Comment (nashton) [reply syncctl.py:313]: Done

File: .github/workflows/pytest.yml (hunk 5 above)
Comment (nashton) [reply .github/workflows/pytest.yml:27]: And now I added the requirements.txt, as the script has dependencies 😀
I think it is fine to keep the check, as it is a generic thing, to make the pipeline usable in other environments.

File: syncctl.py [hunk 6]
```diff
@@ -290,19 +310,49 @@ def mirror_images(manifest: dict, manifest_file: str, incremental: bool) -> None
     os.rename("work/images.tmp", "work/images")
     save_manifest(manifest, manifest_file)
 
-def template_charts(api_versions: list[str], values: dict[str, str]) -> Generator[int, None, None]:
+def template_charts(api_versions: list[str], values: dict[str, str], helm_releases: list[dict], helm_charts: list[Path]) -> Generator[int, None, None]:
+    """ Template helm charts. If no matching helm release is found, supplied values will be used for templating
+
+    Args:
+        api_versions: Limit to specific api versions
```
Comment (jacobcrawford) [inline syncctl.py:317]: api versions of what?

File: syncctl.py [hunk 7]
```diff
@@ -290,19 +310,49 @@ def mirror_images(manifest: dict, manifest_file: str, incremental: bool) -> None
...
+        values: Values to use for templating the charts
```
Comment (jacobcrawford) [inline syncctl.py:318]: Is the mapping chart-name -> values, or how is the dict constructed? How can a string be yaml input to a helm chart?

File: syncctl.py [hunk 8]
```diff
@@ -290,19 +310,49 @@ def mirror_images(manifest: dict, manifest_file: str, incremental: bool) -> None
...
+        helm_releases: Helm releases with values to use for templating the charts
+        helm_charts: Helm charts to template
+    Return: Generator for templated helm charts
```
Comment (jacobcrawford) [inline syncctl.py:321]: The generator is returning an int, how can this be helm charts?

**Review by jacobcrawford — CHANGES_REQUESTED** [review]
a few questions to the types

File: syncctl.py (hunk 6 above)
Comment (nashton) [reply syncctl.py:317]: Not sure. It is for the helm template command, but not something we use. I believe it is out of scope for this PR.

File: syncctl.py (hunk 8 above)
Comment (nashton) [reply syncctl.py:321]: The type hinting seems to be wrong, as it is actually bytes, which then is retyped as a string. I considered it out of scope, as it is code I have not changed in this PR. I fixed the return type of template_chart and template_charts.

File: syncctl.py [hunk 9]
```diff
@@ -172,16 +172,15 @@ def template_flux(root_dir: str, git_repo: str, dir: str, git_repos: dict = coll
...
-
-    manifests = template_flux("work/flux", config["flux_repository"].get("flux_repository", config["flux_repository"]["repository"]), "work/flux/" + config["flux_repository"]["entrypoint"])
+def get_helm_releases(config: dict, manifests: str) -> tuple[dict, list[dict]]:
+    """ Get helm releases and helm repositories from supplied manifests
 
+    Args:
+        config: Config where external helm repositories can be specified
+        manifests: Manifests to get helm releases and repos from
+    Return:
+        tuple with list of helm repositories and list of helm charts
```
[the remark this answers is not in the document]
Comment (nashton) [reply syncctl.py:182]: Removed types from doc, as it is already specified in method signature.

**Review by jacobcrawford — APPROVED** [review]
The tests somehow failed, but I reran them with success! LGTM! 🥳

**Pull request merged**